### ISSUE TITLE
NoMethodError: undefined method `[]' for nil:NilClass in RSpec #create and #update tests (and real-world actions)

### ISSUE THREAD

**Issue by moveson** [graphiti-api/graphiti#28]
I'm using the suggested templates for a basic RSpec tests, and I've got #show, #index, and #destroy working. But for #create, I'm getting `NoMethodError: undefined method '[]' for nil:NilClass`. My test looks like this:

```
require 'rails_helper'

RSpec.describe "courses#create", type: :request do
  subject(:make_request) { jsonapi_post "/api/v2/courses", payload }

  describe 'basic create' do
    let(:payload) do
      {
        data: {
          type: 'courses',
          attributes: {
            name: 'Test Course',
            description: 'A test course',
            nextStartTime: '2018-07-01 06:00:00'
          }
        }
      }
    end

    before { login_as_admin }

    it 'creates the resource' do
      expect { make_request }.to change { Course.count }.by(1)
      course = Course.last

      assert_payload(:course, course, json_item)
    end
  end
end
```
The test fails with:
```
NoMethodError: undefined method `[]' for nil:NilClass
./config/initializers/wrap_parameters.rb:17:in `process_action'
./spec/controllers/api/v2/courses/create_spec.rb:4:in `block (2 levels) in <main>'
./spec/controllers/api/v2/courses/create_spec.rb:23:in `block (4 levels) in <main>'
./spec/controllers/api/v2/courses/create_spec.rb:23:in `block (3 levels) in <main>'
-e:1:in `<main>'
```
The RSpec log indicates:
```
Processing by Api::V2::CoursesController#create as HTML
  Parameters: {"_jsonapi"=>{"data"=>{"type"=>"courses", "attributes"=>{"name"=>"Test Course", "description"=>"A test course", "next_start_time"=>"2018-07-01 06:00:00"}}}}
Completed 401 Unauthorized in 1ms (ActiveRecord: 0.0ms)
```
I'm not sure why it returns 401, as it's throwing the NoMethodError before any response is generated, but you can see that the params seem to be correct.

I have the same issue with my #update test. 

This is not the same issue as described in #16, as my jsonapi_compliable gem is up to date (0.11.1).
Using Rails 5.1.5 and RSpec 3.7.

**Comment by moveson** [comment]
I've just verified that the same problem happens when posting from YARC:
```
Started POST "/api/v2/courses" for 127.0.0.1 at 2018-02-28 00:34:31 -0700
Creating scope :standard_includes. Overwriting existing method Course.standard_includes.
Processing by Api::V2::CoursesController#create as HTML
  Parameters: {"_jsonapi"=>{"name"=>"JSONAPI-Suite", "description"=>"First course created through the new v2 api"}}
  User Load (0.9ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["id", 3], ["LIMIT", 1]]
Can't verify CSRF token authenticity.
   (0.3ms)  BEGIN
  FriendlyId::Slug Load (2.0ms)  SELECT  "friendly_id_slugs".* FROM "friendly_id_slugs" WHERE "friendly_id_slugs"."sluggable_id" = $1 AND "friendly_id_slugs"."sluggable_type" = $2 ORDER BY "friendly_id_slugs".id DESC LIMIT $3  [["sluggable_id", 3], ["sluggable_type", "User"], ["LIMIT", 1]]
   (0.4ms)  COMMIT
ERROR: NoMethodError: undefined method `[]' for nil:NilClass
config/initializers/wrap_parameters.rb:17:in `process_action'
[active_model_serializers] Rendered ActiveModel::Serializer::Null with Hash (0.16ms)
Completed 500 Internal Server Error in 48ms (Views: 29.6ms | ActiveRecord: 3.6ms)
```

**Comment by moveson** [comment]
It appears that the [:data] attribute is being stripped off somewhere in the stack. Here's a backtrace, together with the value of `@payload` inside the constructor of JsonapiCompliable::Deserializer:
```
"Payload is: {\"name\"=>\"JSONAPI-Suite\", \"description\"=>\"First course created through the new v2 api\"}"
ERROR: NoMethodError: undefined method `[]' for nil:NilClass
jsonapi_compliable (0.11.1) lib/jsonapi_compliable/deserializer.rb:64:in `id'
jsonapi_compliable (0.11.1) lib/jsonapi_compliable/deserializer.rb:70:in `block in attributes'
jsonapi_compliable (0.11.1) lib/jsonapi_compliable/deserializer.rb:69:in `tap'
jsonapi_compliable (0.11.1) lib/jsonapi_compliable/deserializer.rb:69:in `attributes'
strong_resources (0.6.3) lib/strong_resources/controller/mixin.rb:17:in `apply_strong_params'
activesupport (5.1.5) lib/active_support/callbacks.rb:413:in `block in make_lambda'
activesupport (5.1.5) lib/active_support/callbacks.rb:178:in `block (2 levels) in halting_and_conditional'
actionpack (5.1.5) lib/abstract_controller/callbacks.rb:12:in `block (2 levels) in <module:Callbacks>'
activesupport (5.1.5) lib/active_support/callbacks.rb:179:in `block in halting_and_conditional'
activesupport (5.1.5) lib/active_support/callbacks.rb:507:in `block in invoke_before'
activesupport (5.1.5) lib/active_support/callbacks.rb:507:in `each'
activesupport (5.1.5) lib/active_support/callbacks.rb:507:in `invoke_before'
activesupport (5.1.5) lib/active_support/callbacks.rb:106:in `block in run_callbacks'
jsonapi_compliable (0.11.1) lib/jsonapi_compliable/base.rb:134:in `block in wrap_context'
jsonapi_compliable (0.11.1) lib/jsonapi_compliable/resource.rb:417:in `block in with_context'
jsonapi_compliable (0.11.1) lib/jsonapi_compliable.rb:62:in `with_context'
jsonapi_compliable (0.11.1) lib/jsonapi_compliable/resource.rb:416:in `with_context'
jsonapi_compliable (0.11.1) lib/jsonapi_compliable/base.rb:133:in `wrap_context'
activesupport (5.1.5) lib/active_support/callbacks.rb:117:in `block in run_callbacks'
activesupport (5.1.5) lib/active_support/callbacks.rb:135:in `run_callbacks'
actionpack (5.1.5) lib/abstract_controller/callbacks.rb:19:in `process_action'
actionpack (5.1.5) lib/action_controller/metal/rescue.rb:20:in `process_action'
actionpack (5.1.5) lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
activesupport (5.1.5) lib/active_support/notifications.rb:166:in `block in instrument'
activesupport (5.1.5) lib/active_support/notifications/instrumenter.rb:21:in `instrument'
activesupport (5.1.5) lib/active_support/notifications.rb:166:in `instrument'
actionpack (5.1.5) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
actionpack (5.1.5) lib/action_controller/metal/params_wrapper.rb:252:in `process_action'
activerecord (5.1.5) lib/active_record/railties/controller_runtime.rb:22:in `process_action'
config/initializers/wrap_parameters.rb:17:in `process_action'
actionpack (5.1.5) lib/abstract_controller/base.rb:124:in `process'
actionview (5.1.5) lib/action_view/rendering.rb:30:in `process'
actionpack (5.1.5) lib/action_controller/metal.rb:189:in `dispatch'
actionpack (5.1.5) lib/action_controller/metal.rb:253:in `dispatch'
actionpack (5.1.5) lib/action_dispatch/routing/route_set.rb:49:in `dispatch'
actionpack (5.1.5) lib/action_dispatch/routing/route_set.rb:31:in `serve'
actionpack (5.1.5) lib/action_dispatch/journey/router.rb:50:in `block in serve'
actionpack (5.1.5) lib/action_dispatch/journey/router.rb:33:in `each'
actionpack (5.1.5) lib/action_dispatch/journey/router.rb:33:in `serve'
actionpack (5.1.5) lib/action_dispatch/routing/route_set.rb:844:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
bullet (5.7.3) lib/bullet/rack.rb:10:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/rack/agent_hooks.rb:30:in `traced_call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/rack/browser_monitoring.rb:32:in `traced_call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
warden (1.2.7) lib/warden/manager.rb:36:in `block in call'
warden (1.2.7) lib/warden/manager.rb:35:in `catch'
warden (1.2.7) lib/warden/manager.rb:35:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
rack (2.0.4) lib/rack/etag.rb:25:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
rack (2.0.4) lib/rack/conditional_get.rb:38:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
rack (2.0.4) lib/rack/head.rb:12:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
rack (2.0.4) lib/rack/session/abstract/id.rb:232:in `context'
rack (2.0.4) lib/rack/session/abstract/id.rb:226:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/cookies.rb:613:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
activerecord (5.1.5) lib/active_record/migration.rb:556:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/callbacks.rb:26:in `block in call'
activesupport (5.1.5) lib/active_support/callbacks.rb:97:in `run_callbacks'
actionpack (5.1.5) lib/action_dispatch/middleware/callbacks.rb:24:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/executor.rb:12:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
better_errors (2.4.0) lib/better_errors/middleware.rb:84:in `protected_app_call'
better_errors (2.4.0) lib/better_errors/middleware.rb:79:in `better_errors_call'
better_errors (2.4.0) lib/better_errors/middleware.rb:57:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/debug_exceptions.rb:59:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
web-console (3.5.1) lib/web_console/middleware.rb:135:in `call_app'
web-console (3.5.1) lib/web_console/middleware.rb:28:in `block in call'
web-console (3.5.1) lib/web_console/middleware.rb:18:in `catch'
web-console (3.5.1) lib/web_console/middleware.rb:18:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
railties (5.1.5) lib/rails/rack/logger.rb:36:in `call_app'
railties (5.1.5) lib/rails/rack/logger.rb:24:in `block in call'
activesupport (5.1.5) lib/active_support/tagged_logging.rb:69:in `block in tagged'
activesupport (5.1.5) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (5.1.5) lib/active_support/tagged_logging.rb:69:in `tagged'
railties (5.1.5) lib/rails/rack/logger.rb:24:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
sprockets-rails (3.2.1) lib/sprockets/rails/quiet_assets.rb:13:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/request_id.rb:25:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
rack (2.0.4) lib/rack/method_override.rb:22:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
rack (2.0.4) lib/rack/runtime.rb:22:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
activesupport (5.1.5) lib/active_support/cache/strategy/local_cache_middleware.rb:27:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/executor.rb:12:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
actionpack (5.1.5) lib/action_dispatch/middleware/static.rb:125:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
rack (2.0.4) lib/rack/sendfile.rb:111:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
webpacker (3.2.2) lib/webpacker/dev_server_proxy.rb:18:in `perform_request'
rack-proxy (0.6.3) lib/rack/proxy.rb:57:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
railties (5.1.5) lib/rails/engine.rb:522:in `call'
newrelic_rpm (4.8.0.341) lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
puma (3.11.2) lib/puma/configuration.rb:225:in `call'
puma (3.11.2) lib/puma/server.rb:624:in `handle_request'
puma (3.11.2) lib/puma/server.rb:438:in `process_client'
puma (3.11.2) lib/puma/server.rb:302:in `block in run'
puma (3.11.2) lib/puma/thread_pool.rb:120:in `block in spawn_thread'
```

**Comment by moveson** [comment]
Further investigation indicates that this is occurring because of my StrongResources configuration:
```
StrongResources.configure do
  strong_resource :course do
    attribute :name, :string
    attribute :description, :string
    attribute :next_start_time, :datetime
  end
end
```
This results in `type_for_param(:datetime)` being called, which causes an error. If :datetime is not to be supported, this should be made clear in the documentation and available types should be restricted in some way.

**Comment by richmolj** [comment]
Hey @moveson - sorry for the headache! But looks like you found the issue.

I've updated the [documentation for SR](https://jsonapi-suite.github.io/strong_resources) calling out built-in types and datetimes. [According to the swagger specification](https://swagger.io/specification/#data-types-13) datetimes are actually specially-formatted strings. I'd love to see support for those specially-formatted strings, but at the moment we only support the more-generic `string`. Keep in mind the bottom of the docs also shows how to define your own custom types.

I'd love to raise a better error here so this would have been more obvious. I've made a note to myself to address this, or if you get a chance to PR seems like an easy win.

Thanks for raising this! ❤️
